active/0024--message-validation-and-transform.md

```diff
@@ -0,0 +1,162 @@
+# Message validation and transform
+
+## Changelog
+
+* 2020-10-21: @zmstone Initial draft
+
+## Abstract
+
+A new feature for EMQX v5 to validate, transform, and filter MQTT messages.
+
+## Motivation
+
+### Data validation
+
+EMQX has no native support for data validation.
+
+Technically, by using EMQX rules engine, it is possible to validate the incoming
+MQTT messages before sending the message off to data bridges.
+
+For instance, one can write an rules engine SQL like below to
+drop messages which are not avro encoded and also when the 'val' field is not greater than 10.
+
+```
+SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
+```
+
+However, currently rules are only used in data bridges, but it does not interfere
+MQTT message pub/sub.
+For example, if there is subscriber subscribing to `t/#` it would still receive the avro encoded message.
+
+If we also want to stop the message being published, then there would be a need for a message republish action,
+and the publisher and subscriber will likely be forced to use different topics.
+e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.
+
+This can work, but not easy to use.
```

Comment:
Random thought, you probably have already pondered it. We could introduce something like _rewrite_ action in the Rule Engine and grant it an ability to erase the original message and replace it with one or more other messages, or even terminate the connection if configured accordingly.

Reply:
Can maybe put this whole new thing to `rule_engine.validation_and_transform`.
However mixing it with existing actions would mean
1. Actions are managed using different APIs, it's probably not the best way to manage validation/transformation
2. Rules lack of ordering, if we are to force the such actions to be positioned in front of the list then it's effectively a separate list.

Reply:
Very good points. A couple of questions / thoughts.

> Actions are managed using different APIs

Q: Different from which?

> Rules lack of ordering, if we are to force the such actions to be positioned in front of the list

I can very well stray away from the initial requirements, but assuming this feature is for some advanced user, then it's easy to imagine that they want to control the order as well, i.e.: 1) log and/or archive in some remote system the message completely as it is, 2) validate / transform it, 3) push through some bridge. However this starts to sound more like _pipelines_ honestly, I'm not sure if we are ready to go that far.

Anyway, I'm just exploring / expressing design ideas, please feel free to ignore them.

Reply:
Actions have a `actions` config root, and they are so far dedicated to data integration, and managed using the `/actions` HTTP endpoint. It's would be odd if we create a `/validations` but put things under the `actions` space.

Actions can have order when pipelined and used in rule, but rules do not have order and it would be hard to define.
To support data validation as a rule, it would imply to create a singleton rule and chained (ordered actions)

Reply:
> To support data validation as a rule, it would imply to create a singleton rule and chained (ordered actions)

Do you imagine here a scenario where a customer wants that "ALL incoming messages should be validated / filtered / transformed first", before even hitting the broker (e.g. routing / hooks)? Then I agree, it sounds impossible to achieve on the data integration side alone. As a sidenote: that sounds a lot like ingress bridge + rules but for the MQTT clients themselves (i.e. not for some remote system).

However, if hitting the broker is not an issue then I don't quite understand why a singleton rule is a must. Again, a sufficiently advanced customer may have a lot of different clients with different patterns, which may require different validation / transformation applied. Thus, different pipelines.

But I agree that ordering between rules is totally unnecessary design complexity.